File: 99.practice/hashmap.cpp
#include<iostream>
#include<vector>
#include<unordered_map>

using namespace std;

void set1(){
unordered_map<string,int> map;
map["IND"] = 1000;
map["USA"] = 100;
map["PK"] = 990;
map["UK"] = 109;
map["NEP"] = 1099;

if(map.find("IND") != map.end())
{
    cout<< map["IND"] << endl;
}
for (auto key:map)
{
    cout << key.first << ":" << key.second << endl;
}
cout<<"------------\n";
for (pair<string,int> key:map)
{
    cout << key.first << ":" << key.second << endl;
}
}

void freqMap(string str)
{
    unordered_map<char,int> map;
    for(char ch: str)map[ch]++;

    for(auto m :map)
    {
        pair<char,int> p = {'e',10};
        // map.insert(p);

        cout<<m.first<<": "<<m.second;
    cout<<endl;
    }
}

void freqMap2(string str)
{
    unordered_map<char,vector<int>> map;
    
    for(int i=0;i<str.length();i++)
    map[str[i]].push_back(i);

    for (auto key:map)
{
    cout << key.first;
    cout<<":";

    for(int d:key.second)
    {
         cout<<d<<" ";
    }
    cout<<endl;
}


}


 

int main()
{
    // freqMap2("SAHITHHHSSSIIAARRTTTTTTAAA");
    freqMap("SAHITHHHSSSIIAARRTTTTTTAAA");
    // set1();
    return 0;
}

File: 99.practice/hashmap.java
import java.util.HashMap;
import java.util.ArrayList;

public class hashmap {

    public static void set1(){
    HashMap<String,Integer> map = new HashMap<>();
        map.put("IND",10000);
        map.put("USA", 1000);
        map.put("NEP", 100);
        map.put("UK", 990);
        map.put("IND", 1000000);
        
        // System.out.println(map.get("IND"));
        // System.out.println(map.getOrDefault("iND",0));
        // System.out.println(map.keySet()); //O(n)
       
        ArrayList<String> keys = new ArrayList<>(map.keySet());
        for(String s:keys)
        {
            // if(s == "IND")
            // {
            //     map.put(s, 000);
            // }
            System.out.println(s + ": " + map.get(s));
        }

        System.out.println( map.containsKey("IND"));
        // System.out.println(map);
    }


    public static void freqMap(String str)
    { 
    HashMap<Character,Integer> map = new HashMap<>();
    for(int i=0;i<str.length();i++)
    {
        char ch = str.charAt(i);
        map.put(ch, map.getOrDefault(ch, 0)+1);

        // if(map.containsKey(ch))
        // {
        //     map.put(ch, map.get(ch)+1);
        // }else{
        //     map.put(ch, 1);
        // }

        // dont use nullptr exception
        // map.put(ch, map.get(ch)+1);
    }
    System.out.println(map);
    }

    public static void freqMap2(String str){
        HashMap<Character,ArrayList<Integer>> map = new HashMap<>();
        for(int i=0;i<str.length();i++)
        {
            char ch = str.charAt(i);
            if(map.containsKey(ch))
            {
                map.get(ch).add(i);
            }else{
                map.put(ch, new ArrayList<>());
            }

        }
        // System.out.println(map);
        ArrayList<Character> keys = new ArrayList<>(map.keySet());
        for(Character s:keys)
        {
            System.out.println(s + ":" + map.get(s));
            System.out.println();
        }
    }

    public static void main(String args[]) {
        // set1();
        // freqMap("SAHITHHHSSSIIAARRTTTTTTAAA");
        freqMap2("SAHITHHHSSSIIAARRTTTTTTAAA");
    }

}
